### ISSUE TITLE
Duplicate corporate body in contribution list 

### ISSUE THREAD

**Issue by jschnasse** [hbz/lobid-resources#933]
Where is the second occurrence of http://d-nb.info/gnd/2116834-9 coming from? As reported by client it does occur only once in MAB source. 

https://lobid.org/resources/HT019838800
http://lobid.org/hbz01/HT019838800

https://lobid.org/resources/HT019837193
http://lobid.org/hbz01/HT019837193

https://lobid.org/resources/HT019826502
http://lobid.org/hbz01/HT019826502

https://lobid.org/resources/HT019826031
http://lobid.org/hbz01/HT019826031

https://lobid.org/resources/HT019825784
http://lobid.org/hbz01/HT019825784

**Comment by dr0i** [comment]
These have in common: a double entry of `200b1.4.`According to our morph this will then fires two times. Fix is to fire when the entity (here: `200`) hits the `?????.9` (the gnd ID) which occurs only once.

**Comment by dr0i** [comment]
Deployed the mentioned resources into staging, see e.g. https://stage.lobid.org/resources/HT019826031.

**Comment by jschnasse** [comment]
+1 this looks much more logical - hm - until we hit a title that defines a contributor type for contributor without GND-ID

**Comment by dr0i** [comment]
Indeed.

**Comment by dr0i** [comment]
Going to deploy this. Needs complete reindexing for all resources => this will be done after complete ETL and after the manual index switch during the course of monday.

**Comment by dr0i** [comment]
As discussed offline with @acka47 the former behaviour was indeed correct: it is a contributor with two roles (also these roles were matched to the default ("Contributor"). In lobid we decided to denote explicitly one contributor with one role. So in this case it would be two times the same contributor, only differing in role.
However, in the frontend we just show under "Mitwirkende" the name of the contributor so that the same name appears two times. If that feels weired, it is up to the rendering to filter to show only one.

**Comment by acka47** [comment]
Generally, it is expected behaviour that we have two `Contribution` objects with the same entity if subfield `4` for a contributor (1--s oder 2--s) is repeated. See #868 for background and especially https://github.com/hbz/lobid-resources/issues/868#issuecomment-405514382 with a link to an email in the Bibframe list where Ray Denenberg says that this is how it should be done (instead of having one object with both roles listed in an array).

It obviously does not make sense when the roles are the same as in this example (the role `ctb` is used twice). This is because we currently haven't covered both `pat` and `isb` in the morph, see [lines 1011-1055 of the current morph](https://github.com/hbz/lobid-resources/blob/4dd00ad21145f6c1793266fef90f506e2a3e491d/src/main/resources/morph-hbz01-to-lobid.xml#L1011-L1055).

Solution: We will not make this one object for http://d-nb.info/gnd/2116834-9 but add the relator codes `pat` and `isb` to the morph and the labels.json. I will open another PR.

**Comment by acka47** [comment]
I now opened a PR for this, see https://github.com/hbz/lobid-resources/pull/935.

However, I am expecting some problems with customers regarding the labels of the relators. Currently we have the following:

- `pat` ("patron") has the label "Im Auftrag von" (I liked it better then "Förderer"/"gefördert von" or "Auftraggeber")
- `isb` has "Herausgeber/in" which is also – correctly – used for the English "editor" (`edt`)

Customers will definitely have a problem with "Herausgeber/in" for `isb`. @jschnasse, we would be happy if customers suggested a better label.

**Comment by dr0i** [comment]
Looks good especially the frontend discriminates the two contributor.

**Comment by dr0i** [comment]
Assigning @jschnasse to receive finishing +1.

**Comment by jschnasse** [comment]
+1 (it is already on production)
> This is because we currently haven't covered both pat and isb in the morph, see lines 1011-1055 of the current morph.

Follow up question: How many roles are not covered in your morph? - And is it possible to cover all possible roles, define labels etc.?

**Comment by acka47** [comment]
> How many roles are not covered in your morph? - And is it possible to cover all possible roles, define labels etc.?

Good question. Christoph did not implement all possible roles but only those that were in the data  – or at least those that appear more frequently – because it is quite tedious to implement.

Generally, there exist 268 different relator codes, see http://id.loc.gov/vocabulary/relators.html. In the morph, roles are handled separately for persons (see [lines 812 to 990](https://github.com/hbz/lobid-resources/blob/d6395ab8b57f862136f21e36791311591ed87bc7/src/main/resources/morph-hbz01-to-lobid.xml#L812-L990) of the current morph) and corporate bodies (see [lines 1011 to 1061](https://github.com/hbz/lobid-resources/blob/d6395ab8b57f862136f21e36791311591ed87bc7/src/main/resources/morph-hbz01-to-lobid.xml#L1011-L1061)).

According to https://www.loc.gov/marc/relators/relaterm.html most relators can apply to corporate bodies as well as to persons. The coverage currently is:
- 15 roles for corporate bodies
- 54 roles for persons (as well as an additional two non-MARC roles)

I know I asked Christoph whether we can not implement a general approach in the moprh for mapping the three-character strings in the data to the MARC relator URI (e.g. `dgs` --> `http://id.loc.gov/vocabulary/relators/dgs`). This only works where the codes are actually existent, i.e. for the records complying to RDA. However there are a lot of additional mappings from earlier used strings to relatores (e.g. `<regexp match="Vorwort|^win$" format="http://id.loc.gov/vocabulary/relators/win"/>
`). Maybe @dr0i can have a look into mapping every code to the URI. 

However, we will also have to add them to the `labels.json` which can't be done automatically as we have to find a German label for each code. (I guess this is the main reason why not all codes are covered.)

My suggestion: We run some statistics on the use of MARC relator codes and add those that are not yet covered. What do you think, @dr0i and @jschnasse ?

**Comment by jschnasse** [comment]
> as we have to find a German label for each code
Is there really no community activity for translating all relator codes to german lang? Something reusable we can use as default? 
If there is nothing, maybe we (together with our customers) should provide and publish a first attempt... Could it make sense to set up a wiki page or something to gather labels for all 268 roles?

**Comment by acka47** [comment]
I also thought about it. I will ask colleages at the DNB whether something like this is already planned or even running. Otherwise we might initiate it in the KIM context. I will think about the best approach.

**Comment by acka47** [comment]
As there are German labels in the source data, there obviously already is a translation list. So I did some research.

I think the terms are based on the "Arbeitshilfe AH-017" that is currrently available in version 1.7 (and of course as PDF) see https://wiki.dnb.de/download/attachments/106042227/AH-017.pdf. 

There is also a VuFind config file at https://github.com/vufind-org/vufind/blob/master/config/vufind/author-classification.ini#L312-L591 based on AH-017 and with gendered labels.

**Comment by dr0i** [comment]
Nice catch @acka47 for Vufind's repo. Provides 279 roles while @jschnasse mentions 268, so it seems to be a complete cover. A correspondig labels.json could be automatically generated. Also, if, desired, the vufind config could be harvested automatically to have updates (since the last update was of 2016 this may be unnecessary).

**Comment by dr0i** [comment]
Should I go to implement this?

**Comment by acka47** [comment]
@dr0i I think this is not urgent to implement. I opened a separate issue at #941 . Closing this one.
